### PR TITLE
[ML] Adds exponent aggregator to the inference model definition

### LINKED ISSUE #1372: [ML] Inference models for MSLE regression trees should have exp aggregators
`CMsle::transform` computes an exponent  of the prediction before writing out the result. This exponentiation is not reflected in the inference model definition, hence the prediction results are off by a huge margin.

### PR DESCRIPTION
**Pull request by valeriy42**
This PR adds exponent aggregator required for the regression models trained with MSLE objective function.

Closes #1372

### CODE REVIEW

File: lib/api/CBoostedTreeInferenceModelBuilder.cc [hunk 1]
```diff
@@ -140,12 +140,20 @@ CRegressionInferenceModelBuilder::CRegressionInferenceModelBuilder(const TStrVec
...
     this->definition().trainedModel()->targetType(CTrainedModel::ETargetType::E_Regression);
 }
 
 void CRegressionInferenceModelBuilder::setAggregateOutput(CEnsemble* ensemble) const {
-    ensemble->aggregateOutput(std::make_unique<CWeightedSum>(ensemble->size(), 1.0));
+    if (m_LossType == TLossType::E_MsleRegression) {
+        ensemble->aggregateOutput(std::make_unique<CExponent>(ensemble->size(), 1.0));
+    } else {
+        ensemble->aggregateOutput(std::make_unique<CWeightedSum>(ensemble->size(), 1.0));
+    }
```
Comment (tveasey) [inline lib/api/CBoostedTreeInferenceModelBuilder.cc:156]: My inclination would be to make this a switch: it avoids us forgetting this in future if we add a new regression type.

File: include/api/CInferenceModelDefinition.h [hunk 2]
```diff
@@ -112,6 +112,27 @@ class API_EXPORT CLogisticRegression final : public CAggregateOutput {
     TDoubleVec m_Weights;
 };
 
+//! Allows to use logistic regression aggregation.
+//!
+//! Given a weights vector $\vec{w}$ as a parameter and an output vector from the ensemble $\vec{x}$,
+//! it computes the logistic regression function \f$1/(1 + \exp(-\vec{w}^T \vec{x}))\f$.
```
Comment (tveasey) [inline include/api/CInferenceModelDefinition.h:118]: This comment looks wrong. The forest is predicting `sum_i{ leaf_i } = log(1 + x)`, so for the weighted case the aggregator would be `exp(w^t x) - 1` no?

Comment (tveasey) [reply include/api/CInferenceModelDefinition.h:118]: In fact, thinking about it, 1 is the offset is a parameter which we also need to supply.

Comment (valeriy42) [reply include/api/CInferenceModelDefinition.h:118]: You are right, the aggregator computes `exp(w^t x)` cf. `CMsle::transform`.

File: lib/api/CBoostedTreeInferenceModelBuilder.cc (hunk 1 above)
Comment (valeriy42) [reply lib/api/CBoostedTreeInferenceModelBuilder.cc:164]: I added a switch and `LOG_ERROR` if we unexpectedly encounter a classification function.

**Comment by valeriy42** [comment]
Thank you @tveasey for the review. I addressed your comments. It would be great if you could take another look.

File: include/api/CInferenceModelDefinition.h (hunk 2 above)
Comment (tveasey) [reply include/api/CInferenceModelDefinition.h]: Ok great.

and the comment on the offset was me forgetting what we're actually doing. For posterity, we compute leaf values by `argmin_x (log(offset + exp(prediction)) - log(offset + actual))^2`, i.e. the important part is that prediction is exponentiated so the forest is effectively predicting logs. The offset just governs how we penalise relative errors between the prediction actuals at the scale of offset.

File: lib/api/CBoostedTreeInferenceModelBuilder.cc (hunk 1 above)
Comment (tveasey) [reply lib/api/CBoostedTreeInferenceModelBuilder.cc:164]: Thanks!

**Comment by valeriy42** [comment]
retest

**Pull request merged**